lib/utf8_sanitizer.rb

```diff
@@ -9,7 +9,9 @@ def initialize(app)
   def call(env)
     parser = RackRequestParser.new(Rack::Request.new(env))
 
-    if invalid_strings(parser.values_to_check)
+    if has_invalid_strings(parser.values_to_check) ||
+       params_have_null_byte?(parser.request.params) ||
+       has_null_byte?(parser.request.body)
```

Comment:
Couple things here:

- Would it be simpler / faster to just check the raw values instead, rather than parsing and walking the data form?
- I think `parser.request.params` is already accounting for the body? ([source](https://github.com/rack/rack/blob/d0c6efc666ede26768f33935f00530629690369a/lib/rack/request.rb#L459-L464))
- In my testing `parser.request.body` is a `StringIO`, not a string, so we'd need to `.read` it first?

```suggestion
       parser.request.query_string.include?("\x00") ||
       parser.request.body.read.include?("\x00")
```

(could be moved to separate methods)

Reply:
That's a good point. But what if somebody encodes a null byte in JSON? The plain string JSON itself doesn't contain a null byte, but the corresponding `params` would...

```ruby
payload = { "a" => "\x00" }
payload["a"].include?("\x00")
# => true
str = payload.to_json
# => "{\"a\":\"\\u0000\"}"
str.include?("\x00")
# => false
```

Reply:
Yeah, that's fair. I spent a bit of time last week digging into this as well, and it quickly became obvious that it's more complicated than I'd expected with managing encoded forms. If those are at least predictable, it might be worth digging into for performance' sake, but on the other hand it looks like it might actually be the opposite of what I'd hoped in being "simpler".

Reply:
My thinking is, this reflects what will happens when the params goes through the Rails app, so this is our best way to block that? And most of the time, params are small so I don't think it's that scary?

Reply:
Any thoughts on the second item from the original comment?

> * I think `parser.request.params` is already accounting for the body? ([source](https://github.com/rack/rack/blob/d0c6efc666ede26768f33935f00530629690369a/lib/rack/request.rb#L459-L464))

Reply:
> Any thoughts on the second item from the original comment?
> 
> > * I think `parser.request.params` is already accounting for the body? ([source](https://github.com/rack/rack/blob/d0c6efc666ede26768f33935f00530629690369a/lib/rack/request.rb#L459-L464))

I guess that assumes we only care if the body is parsed as parameters. Not sure if we're trying to be broader and catch any null bytes in any body.

Reply:
I think it's OK if the body contains, our problem is blocking these from the database, and as far as I know we never pass the body to the database before parsing it

Reply:
I can try removing the check for now

Reply:
Removed, simplified in 94fa29697819d0c4b2d01770b4959cf7dfc79490